### ISSUE TITLE
Non-nullable shadow property problem

### ISSUE THREAD

**Issue by markm77** [dotnet/efcore#22352]
Hi there,

If I specify a string property in an entity type class as

```csharp
        public string Test2 { get; set; } = null!;
```

it gets correctly picked up as non-null (```"Test2" TEXT NOT NULL```) in the create statement.

If I specify the same property as a shadow property via

```csharp
            modelBuilder
                .Entity<MyEntity>()
                .Property<string>("Test1");
```

it is nullable (```"Test1" TEXT NULL```) in the create statement.

This seems like a bug or am I doing something wrong?

### Further technical details

EF Core version: 5.0.0-preview.8.20407.4
Database provider: Sqlite
Target framework: .NET Core 3.1 latest (nullable enabled)
Operating system: Windows 10 latest
IDE: Rider latest

**Comment by smitpatel** [comment]
Do you have nullable enable in the code where you are declaring shadow property?

**Comment by markm77** [comment]
Yes, project-wide. I tried with ```string?``` in Fluent and got same result.

**Comment by smitpatel** [comment]
cc: @roji 
Verified that this is bug in latest bits. When computing nullability based on clr type purely, we don't check for the type to have all annotations related to non-nullable reference type.

**Comment by markm77** [comment]
FYI seems like this issue also affects [Owned] properties, e.g. below fields do not map to non-nullable....
```csharp
    [Owned]
    public class OwnedClass
    {
        public string MyString { get; set; }
        public bool MyBool { get; set; }
    }
```

**Comment by ajcvickers** [comment]
Note from triage: fix this for indexer properties in 5.0.

**Comment by roji** [comment]
To the best of my understanding, this is not currently possible, since nullability information for generic type parameters simply isn't available. `typeof(T)` in the above sample always returns string, and unlike with properties/fields/methods there is (currently) no place for additional attributes to hold the nullability info.

See https://github.com/dotnet/csharplang/issues/3587 for a discussion around this (and https://github.com/dotnet/csharplang/issues/3587 for some previous discussion). I've posted https://github.com/dotnet/csharplang/issues/3601#issuecomment-689480937 to provide our case for this, but it does seem rather like a corner case to me, considering the language changes needed to support this.

The workaround for the above is simply to use the fluent API to specify nullability (IsRequired).

**Comment by markm77** [comment]
@roji For what it's worth I'm fine with this if there is a workaround, the [Owned] situation is more of an issue for me as it doesn't seem there's a workaround.

In case it's helpful, I have myself had issues with generic class with nullable parameter X and solved them in the past by having two generic classes - one with X and the other with non-nullable X. Not sure that would work here though as you might have to try/catch the non-nullable case first or something ugly like that since you don't know in advance which the user needs.

**Comment by markm77** [comment]
Following on from my previous comment @roji , can I ask how you actually detect a reference-type is nullable in case of non-Fluent properties case given I thought nullability is a compile-time construct only? No problem if you have no time to answer.

**Comment by roji** [comment]
@markm77 yeah, non-nullable reference types definitely have their limitations and complexity, a result of retrofitting an existing language with such a fundamental new concept.

> Following on from my previous comment @roji , can I ask how you actually detect a reference-type is nullable in case of non-Fluent properties case given I thought nullability is a compile-time construct only?

For properties, fields and methods, nullability is represented via a set of attributes generated by the compiler. The original goal of these attributes are for the compiler itself to know the nullability of APIs when you reference an external assembly, but it's also possible to use reflection to read them. EF Core currently does this in [NonNullableConventionBase](https://github.com/dotnet/efcore/blob/19a65f2fa8abef9a02f5ab45e4941dc0ee128698/src/EFCore/Metadata/Conventions/NonNullableConventionBase.cs) (although certain scenarios with generic entity types aren't supported, it's always about generic :smile: ).

If you're interested in the actual metadata, [see these specs](https://github.com/dotnet/roslyn/blob/master/docs/features/nullable-metadata.md). There's also which would introduce a 1st-class reflection API for reading this metadata.

https://github.com/dotnet/runtime/issues/29723

**Comment by markm77** [comment]
Hey thanks @roji , great info! I see NRT as following the TypeScript model where you allow compile-time and run-time languages to diverge to add new features without modifying the run-time language.
However I do hope at some point the run-time language (CLR) is updated to natively support NRT as this would simplify the whole feature and avoid increasing use of reflection which is not great for performance/debugging/analyzers/etc. Just my thoughts, have a great day!

**Comment by roji** [comment]
Not sure there is anything to do here... For shadow properties specifically, as written above, the runtime Type for generic type parameters lacks NRT information (by design). For non-shadow generic properties, our current NRT metadata decoding indeed doesn't support them because of the complexities involved. However, using the new .NET decoder should fix this - already tracked by #24744.
